[torchbind] Improve IValue custom class API and remove most Capsule stuff

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34933 [torchbind] Enable tests on rocm
* #34906 [torchbind] Add more comprehensive docscrings
* **#34848 [torchbind] Improve IValue custom class API and remove most Capsule stuff**

This makes it so that `IValue(intrusive_ptr<T>)` where `T` inherits from `CustomClassHolder` constructs the proper Object with capsule attribute. It removes the behavior where that constructs a Capsule directly and moves that to `IValue::make_capsule` where it's less of a footgun. Also introduces `toCustomClass<T>()` which will do the inverse, it'll take the full `Object` and pull the custom class out from the capsule, returning the `intrusive_ptr`

Differential Revision: [D20480514](https://our.internmc.facebook.com/intern/diff/D20480514)

Commit:
[torchbind] Improve IValue custom class API and remove most Capsule stuff

[ghstack-poisoned]

Commit:
Update on "[torchbind] Improve IValue custom class API and remove most Capsule stuff"




[ghstack-poisoned]

Commit:
Update on "[torchbind] Improve IValue custom class API and remove most Capsule stuff"




[ghstack-poisoned]

Commit:
Update on "[torchbind] Improve IValue custom class API and remove most Capsule stuff"


Differential Revision: [D20480514](https://our.internmc.facebook.com/intern/diff/D20480514)

[ghstack-poisoned]

Commit:
Update on "[torchbind] Improve IValue custom class API and remove most Capsule stuff"


This makes it so that `IValue(intrusive_ptr<T>)` where `T` inherits from `CustomClassHolder` constructs the proper Object with capsule attribute. It removes the behavior where that constructs a Capsule directly and moves it to `make_capsule`, where it's less of a footgun. Also introduces `toCustomClass<T>()` which will do the inverse, it'll take the full `Object` and pull the custom class out from the capsule, returning the `intrusive_ptr`

Differential Revision: [D20480514](https://our.internmc.facebook.com/intern/diff/D20480514)

[ghstack-poisoned]

Commit:
Update on "[torchbind] Improve IValue custom class API and remove most Capsule stuff"


This makes it so that `IValue(intrusive_ptr<T>)` where `T` inherits from `CustomClassHolder` constructs the proper Object with capsule attribute. It removes the behavior where that constructs a Capsule directly and moves that to `IValue::make_capsule` where it's less of a footgun. Also introduces `toCustomClass<T>()` which will do the inverse, it'll take the full `Object` and pull the custom class out from the capsule, returning the `intrusive_ptr`

Differential Revision: [D20480514](https://our.internmc.facebook.com/intern/diff/D20480514)

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -207,14 +207,25 @@ struct CAFFE2_API IValue final {
+  c10::intrusive_ptr<T> toCustomClass() &&;
+  template <typename T>
+  c10::intrusive_ptr<T> toCustomClass() const &;
+
```
Normally when we add a specific `toX` we also add to the templated `.to<X>`. Does `.to<MyCustomClass>()` work?

jamesr66a replied:
it ends up being `.to<c10::intrusive_ptr<MyCustomClass>>()`. Maybe to normalize it we should kill the need to put `intrusive_ptr` there, I can look into it

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -521,14 +522,40 @@ std::vector<Elem> generic_to(
+  TORCH_CHECK(obj->slots().size() == 1,
+              "Tried to cast IValue to custom class but it did "
+              "not contain a custom class!");
+  // FIXME: this cast isn't actually checked!
```
Options:
1. fix before land
2. name this unsafeToCustomClass until it checks this. Our users will find this method and try to call it.

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -774,10 +801,32 @@ inline IValue::IValue(c10::intrusive_ptr<ivalue::PyObjectHolder> v)
+  auto ivalue_obj = c10::ivalue::Object::create(
+      c10::StrongTypePtr(nullptr, classType), /*num_slots=*/1);
+  ivalue_obj->setAttr(
+      "capsule",
```
setSlot would be faster

zdevito commented on torch/custom_class.h:
```diff
@@ -74,10 +74,8 @@ class class_ {
-      auto capsule = c10::IValue(std::move(genericPtr));
-      auto object = std::move(self.ivalue).toObject();
-      object->setSlot(0, std::move(capsule));
+      auto object = self.ivalue.toObject();
```
nit: std::move was dropped here.

jamesr66a replied:
I don't think `std::move()` was correct here, it was added as a codemod and was moving this IValue out, when the `__init__` method should just modify it in place

zdevito reviewed (approved):
Looks good, thanks!

Commit:
Update on "[torchbind] Improve IValue custom class API and remove most Capsule stuff"


This makes it so that `IValue(intrusive_ptr<T>)` where `T` inherits from `CustomClassHolder` constructs the proper Object with capsule attribute. It removes the behavior where that constructs a Capsule directly and moves that to `IValue::make_capsule` where it's less of a footgun. Also introduces `toCustomClass<T>()` which will do the inverse, it'll take the full `Object` and pull the custom class out from the capsule, returning the `intrusive_ptr`

Differential Revision: [D20480514](https://our.internmc.facebook.com/intern/diff/D20480514)

[ghstack-poisoned]

Commit:
Update on "[torchbind] Improve IValue custom class API and remove most Capsule stuff"


This makes it so that `IValue(intrusive_ptr<T>)` where `T` inherits from `CustomClassHolder` constructs the proper Object with capsule attribute. It removes the behavior where that constructs a Capsule directly and moves that to `IValue::make_capsule` where it's less of a footgun. Also introduces `toCustomClass<T>()` which will do the inverse, it'll take the full `Object` and pull the custom class out from the capsule, returning the `intrusive_ptr`

Differential Revision: [D20480514](https://our.internmc.facebook.com/intern/diff/D20480514)

[ghstack-poisoned]
